File: 3.Ottobre/75.HiddenTreasure-24Ottobre.js
//todo  Dato un array 2D che rappresenta una mappa del fondale oceanico che include un tesoro nascosto e un array con le coordinate ([riga, colonna]) per la prossima immersione della tua ricerca del tesoro, restituisci "Vuoto", "Trovato" o "Recuperato" utilizzando le seguenti regole:

//! L'array 2D specificato conterrà esattamente un tesoro non recuperato, che occuperà più celle.

//! Ogni cella dell'array 2D conterrà uno dei seguenti valori:

//! "-": Nessun tesoro.
//! "O": Una parte del tesoro che non è stata trovata.
//! "X": Una parte del tesoro che è già stata trovata.
//! Se il luogo dell'immersione non contiene tesori, restituisci "Vuoto".

//! Se il luogo dell'immersione trova un tesoro, ma almeno un'altra parte del tesoro rimane non trovata, restituisci "Trovato".

//! Se il luogo dell'immersione trova l'ultima parte non trovata del tesoro, restituisci "Recuperato".

//! Ad esempio, dati:

//! [
//! [ "-", "X"],
//![ "-", "X"],
//! [ "-", "O"]
//! ]
//! E [2, 1] per le coordinate del luogo dell'immersione, restituisci "Recuperato" perché l'immersione ha trovato l'ultima parte non trovata del tesoro.

//? Tempo di esecuzione:
//* Non compeltato 

//?Codice scritto senza utilizzo di ChatGpt

function dive(map, coordinates) {

    let w = coordinates[0]
    let h = coordinates[1]

    let valoreW = map[w][h]

    if(valoreW === "X") return "Recovered";
    if(valoreW === "O") return "Found";
    if(valoreW === "-") return "Empty";

    console.log(valoreW,valoreH)
}

console.log(dive([[ "-", "X"], [ "-", "X"], [ "-", "O"]], [2, 1]))

//?Codice scritto da ChatGpt

function dive(map, coordinates) {
  const [row, col] = coordinates;
  const cell = map[row][col];

  if (cell === "-") return "Empty";

  // conta le parti NON trovate, escludendo la cella colpita
  let remainingO = 0;

  for (let r = 0; r < map.length; r++) {
    for (let c = 0; c < map[r].length; c++) {
      if (!(r === row && c === col) && map[r][c] === "O") {
        remainingO++;
      }
    }
  }

  if (cell === "O" && remainingO === 0) return "Recovered";
  return "Found";
}

//* ✅ Pro

//* Codice chiaro e leggibile

//* Accesso corretto a riga/colonna (map[w][h])

//* Gestisci bene i 3 simboli base (-, O, X)

//! ❌ Contro (importanti)

//! Logica incompleta:
//! "Recovered" non dipende solo dalla cella "X" → devi controllare se esistono ancora "O" nella mappa

//! "Found" è sbagliato: dovrebbe scattare solo se rimangono altre parti non trovate

//! console.log(valoreW,valoreH) → bug (valoreH non esiste)

//! Nessuna gestione di coordinate fuori range

//? Versione Chat Gpt 

//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/75.Hidden_treasure-24Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

"""
Dato un array bidimensionale che rappresenta una mappa del fondale oceanico
contenente un tesoro nascosto, e un array con le coordinate ([riga, colonna])
per la prossima immersione di ricerca, restituisci:

- "Empty"
- "Found"
- "Recovered"

Regole:

"-" = Nessun tesoro.
"O" = Parte del tesoro non ancora trovata.
"X" = Parte del tesoro già trovata.

Se nel punto di immersione non c'è il tesoro → "Empty".
Se viene trovata una parte del tesoro ma ne rimangono altre → "Found".
Se viene trovata l'ultima parte del tesoro → "Recovered".
"""

# =========================================================
# //* Codice scritto senza utilizzo di ChatGPT
# =========================================================

def dive(map, coordinates):
    [row_co, col_co] = coordinates
    cell = map[row_co][col_co]

    if cell == "-":
        return "Empty"

    remaining = 0

    for row in range(len(map)):
        for col in range(len(map[0])):
            if not (row == row_co and col == col_co) and map[row][col] == "O":
                remaining += 1

    if cell == "O" and remaining == 0:
        return "Recovered"

    return "Found"

print(dive([
    ["-", "-", "-"],
    ["X", "O", "X"],
    ["-", "-", "-"]
], [1, 2]))

# =========================================================
# //* Codice scritto da ChatGPT (versione migliorata)
# =========================================================

def dive(map, coordinates):
    row, col = coordinates
    cell = map[row][col]

    if cell == "-":
        return "Empty"

    remaining = sum(
        1
        for r in range(len(map))
        for c in range(len(map[0]))
        if (r, c) != (row, col) and map[r][c] == "O"
    )

    return "Recovered" if cell == "O" and remaining == 0 else "Found"

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Logica corretta.
# - Gestisce tutti i casi richiesti dalla consegna.
# - L'idea di ignorare la cella appena esplorata è corretta.
# - Il codice è semplice da seguire.
# - Buon utilizzo degli indici della matrice.

# //! Contro
# - I ";" finali sono inutili in Python.
# - "map" è il nome di una funzione built-in di Python: meglio evitarlo
#   (ad esempio usare seabed, grid o ocean_map).
# - "remaining" viene incrementata quando basta sapere se esiste almeno un'altra
#   "O": si potrebbe interrompere il ciclo appena trovata.
# - L'espressione:
#       not(row == row_co and col == col_co)
#   è corretta ma poco leggibile rispetto a:
#       (row, col) != (row_co, col_co)

# =========================================================
# //* Consigli
# =========================================================

# - Evita di usare nomi di funzioni built-in come variabili (map, list, dict...).
# - Non usare ";" a fine riga.
# - Quando ti interessa solo sapere se esiste almeno un elemento,
#   valuta di usare "any()" oppure interrompere il ciclo con return.
# - Continua così: la logica è stata costruita molto bene.

# =========================================================
# ⭐ Valutazione
# =========================================================

# Correttezza:      10/10
# Leggibilità:      9/10
# Efficienza:       8.5/10
# Stile Python:     8.5/10
#
# Voto complessivo: 9.2/10
#
# Hai risolto l'esercizio senza scorciatoie e con una logica molto solida.
# Le uniche migliorie riguardano lo stile Python e qualche piccola ottimizzazione,
# non l'algoritmo, che è corretto.